[DR-2784] Add initial Firecloud group support for DUOS integration

okotsopoulos opened the pull request:
https://broadworkbench.atlassian.net/browse/DR-2784

This PR adds support for creating, recording, and retrieving the Firecloud groups which will serve as containers for a DUOS dataset’s authorized users.  Presently, this code is not yet wired in anywhere or accessible to the end user, but I've elected to break up my work with the goal of easier review.

**Changes**
I broke these up into logical commits if you'd prefer to step through them.
- We can now call `IamService` to create or delete a Firecloud managed group.  All of these actions are performed as the TDR service account, which will be the group's sole administrator.
- We store records of our TDR-managed Firecloud groups in new Postgres table `duos_firecloud_group`, which associates these groups with their DUOS dataset ID.
- `DuosService` introduces support for creating or retrieving TDR-managed Firecloud groups, leveraging `DuosDao` to record them.
- Added unit tests throughout.

okotsopoulos committed:
Add new IamService calls for creating and fetching Firecloud groups

okotsopoulos committed:
Add duos_firecloud_group table to Postgres

This is a table mapping DUOS IDs to their Firecloud managed group, and includes audit information -
including the TDR SA which created the group.

okotsopoulos committed:
Introduce DuosDao (insert and retrieve), unit tests, and data model

okotsopoulos committed:
Introduce DuosService (create and/or retrieve) and unit tests

In the process, realized that we may also like to be able to delete Firecloud groups if we can't successfully record them on creation.
Expanded IamService and tests accordingly.

nmalfroy commented on src/main/java/bio/terra/service/auth/iam/IamProviderInterface.java:
```diff
@@ -217,4 +217,17 @@ String getPetToken(AuthenticatedUserRequest userReq, List<String> scopes)
+   * @param groupName name of Firecloud managed group
+   * @return the group's email address
+   */
+  String getGroupEmail(String accessToken, String groupName) throws InterruptedException;
```
Generally, we want to pass in `AuthenticatedUserRequest userReq` over access token (I think that the main reason we don't in `registerUser` is that it's a bit of a special case before the user has been created)

okotsopoulos replied:
I see you figured out my intention in this other comment:

https://github.com/DataBiosphere/jade-data-repo/pull/1361/files/23f8189381f996f7b3acae0bd06007dd2eb88086#r999573134

But I will add top-level comments to the `IamService` calls to indicate that it's the TDR SA who is performing these calls!  I don't want to bake that into the interface since we're generating the SA token upstream.

nmalfroy commented on src/main/java/bio/terra/service/auth/iam/IamProviderInterface.java:
```diff
@@ -217,4 +217,17 @@ String getPetToken(AuthenticatedUserRequest userReq, List<String> scopes)
+   * @param accessToken valid oauth token for the account creating the group
+   * @param groupName name of Firecloud managed group to create
+   */
+  void createGroup(String accessToken, String groupName) throws InterruptedException;
```
Same thing about using an `AuthenticatedUserRequest` object.  Do we want to return something? Like maybe the email?

nmalfroy replied:
Wait, sorry, scratch what I said about AuthenticatedUserRequest...I see that it's meant to be used with a service account token.  I do think we should return something though

nmalfroy replied:
bah, nevermind about the return.  I see that the API doesn't return anything.  Bummer.  Do we need a delete too?  Not sure if we ever plan on implementing that if all instances of a DUOS ID are gone

okotsopoulos replied:
Good questions!

- Void method: yeah, I wish the SAM API behaved differently in this case!  I toyed with making `createGroup` return the group's email though.  In fact, I think I'll make that change to reduce the burden on the caller, as it's clearly not the most intuitive.
- Delete group: I ended up adding this in a later commit!  The way it is presently used is if we successfully create a group but fail to insert it in the DB, we delete it to avoid proliferation of unrecorded groups. https://github.com/DataBiosphere/jade-data-repo/blob/7fcaf145670773d610a26ed56d9cf0db03ccba62/src/main/java/bio/terra/service/duos/DuosService.java#L47-L52

nmalfroy replied:
OMG, giant facepalm...I forgot that I was reviewing commit-by-commit.

nmalfroy reviewed (approved):
Looks great!  I left my comments in just in case others have a similar train of thought...but mainly, feel free to ignore them :) Only thing is I'm not sure if we need the delete. Probably not a blocker for this PR

nmalfroy commented on src/main/resources/db/changesets/20221014_duosfirecloudgroup.yaml:
```diff
@@ -0,0 +1,38 @@
+        - createTable:
+            tableName: duos_firecloud_group
+            columns:
+              - column:
```
general practice: do we want to include an id field?

okotsopoulos replied:
I didn't add one because the DUOS dataset ID is the primary key, and an additional identifier seemed superfluous.  Would you prefer I add one anyway?

nmalfroy replied:
I would.  We do a similar thing with datasets, snapshots, etc. and it can be helpful if we want to ever support a case where there's a rename (not likely but thinking a head and figured it'd be good for consistency)

okotsopoulos replied:
Can do, thanks for the historical context!

nmalfroy replied:
no problem, thanks!

nmalfroy commented on src/main/java/bio/terra/service/duos/DuosDao.java:
```diff
@@ -0,0 +1,98 @@
+      readOnly = true)
+  public DuosFirecloudGroupModel retrieveFirecloudGroup(String duosId) {
+    try {
+      String sql = "SELECT * FROM duos_firecloud_group WHERE duos_id = :duos_id";
```
would you mind explicitly enumerating the columns that get queried?

okotsopoulos replied:
Sure!

nmalfroy commented on src/main/java/bio/terra/service/duos/DuosDao.java:
```diff
@@ -0,0 +1,98 @@
+            .addValue("duos_id", duosId)
+            .addValue("firecloud_group_name", firecloudGroupName)
+            .addValue("firecloud_group_email", firecloudGroupEmail)
+            .addValue("created_by", tdrServiceAccountEmail);
```
We should track created_date here too right?  Or is that populated automatically?

okotsopoulos replied:
Correct, I made `created_date` an auto-generated column.

nmalfroy commented on src/main/java/bio/terra/service/auth/iam/IamProviderInterface.java:
```diff
@@ -230,4 +230,10 @@ String getPetToken(AuthenticatedUserRequest userReq, List<String> scopes)
+   * @param accessToken valid oauth token for the account creating the group
+   * @param groupName name of Firecloud managed group to delete
+   */
+  void deleteGroup(String accessToken, String groupName) throws InterruptedException;
```
haha there it is :)

nmalfroy commented on src/main/java/bio/terra/service/duos/DuosService.java:
```diff
@@ -0,0 +1,76 @@
+    String groupEmail = iamService.getGroupEmail(groupName);
+    logger.info("Successfully created Firecloud group {} for {} users", groupName, duosId);
+
+    boolean inserted = duosDao.insertFirecloudGroup(duosId, groupName, groupEmail);
```
I wonder if to be extra safe, we should put this and the block at line 48 in a try block?

okotsopoulos replied:
That's a good point, especially as we have the fallback behavior to delete the created group if insertion fails.

okotsopoulos replied:
Sharing our Slack convo for posterity:

What I came up with looked suspiciously like something that should be its own step in a flight, given the cleanup behavior:
```
    boolean inserted = false;
    try {
      inserted = duosDao.insertFirecloudGroup(duosId, groupName, groupEmail);
    } catch (Exception ex) {
      // Unexpected exceptions will be thrown after deleting the Firecloud group
      // in the "finally" block.
      throw ex;
    } finally {
      if (!inserted) {
        iamService.deleteGroup(groupName);
        throw new DuosFirecloudGroupInsertException(
            "Firecloud group " + groupName + " was not inserted into the DB and has been deleted");
      }
    }
```
That will be coming soon, in its own ticket: https://broadworkbench.atlassian.net/browse/DR-2787
So at this point, I left a comment to that effect and will lightly refactor to separate group creation from DB insertion when I'm building the flight.

okotsopoulos replied:
And… thanks to @pshapiro4broad 's line of questioning, I was able to streamline this and get our try block after all :)

https://github.com/DataBiosphere/jade-data-repo/pull/1361#discussion_r999929413

nmalfroy commented on src/main/java/bio/terra/service/duos/DuosService.java:
```diff
@@ -0,0 +1,76 @@
+    boolean inserted = duosDao.insertFirecloudGroup(duosId, groupName, groupEmail);
+    if (!inserted) {
+      iamService.deleteGroup(groupName);
+      throw new RuntimeException(
```
I think we want a more specific exception (we generally only use RuntimeException in tests)

nmalfroy reviewed (commented):
Still looks good :). Just a couple of extra comments

okotsopoulos committed:
@nmalfroy PR feedback - consolidate email fetch with group creation, clarify TDR SA's role in group ops

Updated calls and unit tests accordingly.

okotsopoulos committed:
@nmalfroy PR feedback

- Add id field to duos_firecloud_group
- Explicitly enumerate duos_firecloud_group columns queried on retrieval
- Throw specific exception rather than RuntimeException when DB insertion fails
- Add comment flagging desired step breakdown when implementing set / unset snapshot DUOS ID flight

pshapiro4broad commented on src/main/java/bio/terra/service/duos/DuosDao.java:
```diff
@@ -0,0 +1,104 @@
+            .addValue("firecloud_group_email", firecloudGroupEmail)
+            .addValue("created_by", tdrServiceAccountEmail);
+    int rowsAffected = jdbcTemplate.update(sql, params);
+    boolean insertSucceeded = (rowsAffected == 1);
```
Is there a way that this can be `false`; that is, what could cause `rowsAffected` to be `0` but no exception to be thrown?

pshapiro4broad replied:
I see, if `duosId` is not found in the database, no exception is thrown and `update()` returns `0`.

okotsopoulos replied:
Actually, I think your initial question is right on.  If we haven't inserted a row, then something exceptional occurred and would have shown up as a thrown exception: violation of a constraint, etc.  We're better served by allowing the exception to bubble up rather than obfuscating it.

okotsopoulos replied:
Making this change ended up with the added benefit of streamlining error handling at the caller, so thanks!

okotsopoulos commented on src/main/resources/db/changesets/20221014_duosfirecloudgroup.yaml:
```diff
@@ -0,0 +1,45 @@
+                  type: ${uuid_type}
+                  defaultValueComputed: ${uuid_function}
+                  constraints:
+                    primaryKey: true
```
Whoops, this new `id` column should not be a primary key… otherwise we will permit the creation of records with the same DUOS ID. I will update and expand unit tests.

okotsopoulos committed:
Generated duos_firecloud_group.id should not be a pkey

Otherwise, we allow multiple rows written with the same DUOS ID.
Expand unit testing to verify this behavior.

okotsopoulos committed:
@pshapiro4broad idea - bubble up duos_firecloud_group insertion exceptions

Either we insert 1 record, or an exception is thrown.
We now allow the exception to propagate rather than be obscured behind a boolean.
DuosService.createFirecloudGroup fallback behavior on failed insert is streamlined as a result.

snf2ye reviewed (approved):
Thanks for making this PR so easy to go through and for walking me through the bigger picture yesterday! I'm sure there will be things we find as you wire things up, but looks like you (+other reviewers) did a great job considering edge cases and thinking everything through.
